File: linreg.py
import numpy as np

class LinearRegression:
    def __init__(self, lr=0.01, n_iters=1000):
        """
        Initialize the Linear Regression model.
        :param lr: Learning rate for gradient descent.
        :param n_iters: Number of iterations for gradient descent.
        """
        self.lr = lr
        self.n_iters = n_iters
        self.weights = None

    def fit(self, X, y):
        """
        Train the model using gradient descent.
        :param X: Input features (numpy array).
        :param y: Target variable (numpy array).
        """
        n_samples, n_features = X.shape
        # Initialize weights
        self.weights = np.zeros((n_features, 1))

        # Reshape y to a column vector for matrix operations
        y = y.reshape(-1, 1)

        # Gradient Descent loop
        for _ in range(self.n_iters):
            # Calculate predictions (hypothesis: y = Xw + b)
            y_pred = np.dot(X, self.weights)

            # Calculate gradients
            # Gradient of loss w.r.t weights (dw)
            dw = (2 / n_samples) * np.dot(X.T, (y_pred - y))

            # Update weights
            self.weights -= self.lr * dw

    def predict(self, X):
        """
        Predict new values using the trained model.
        :param X: Input features (numpy array).
        :return: Predicted values.
        """
        y_pred = np.dot(X, self.weights)
        return y_pred
    
    def mse(self, X, y):
        y_hat = self.predict(X)
        y = y.reshape(-1,1)
        return np.mean((y_hat-y)**2)

# Example Usage:
if __name__ == '__main__':
    # Generate some synthetic data
    np.random.seed(42)
    n,p = (100,2)
    X = 2 * np.random.rand(n, p+1)
    X[:,0] = 1.0
    true_weights = np.array([[4.0], [3.0], [5.0]])
    y = np.dot(X, true_weights) + np.random.randn(100, 1)

    # Create and train the model
    regressor = LinearRegression(lr=0.01, n_iters=1000)
    regressor.fit(X, y)

    print(X.shape, y.shape)
    # Print the learned parameters
    print(f"Learned weights: {regressor.weights}")
    # Expected values are close to weights: [4.0,3.0,5.0]
    print(f"MSE: {regressor.mse(X,y):.3f}")

File: src/linreg.rs
use ndarray::{Array1, Array2};

pub struct LinearRegression {
    pub lr: f64,
    pub weights: Option<Array1<f64>>, // None before fit - Some after fit
    pub n_iters: i32,
}

impl LinearRegression {
    pub fn new(lr: f64, n_iters: i32) -> Self {
        Self {
            lr,
            weights: None,
            n_iters,
        }
    }

    pub fn fit(&mut self, X: &Array2<f64>, y: &Array1<f64>){
        // Given X and Y, fit a linear regression model
        
        // Get shape of X matrix
        let (n,p) = X.dim();

        self.weights = Some(Array1::zeros(p));

        // Extract weights as 'w' from Option
        // let's work with a local mutable copy
        let mut w = self.weights.take().unwrap();

        // Gradient Descent Loop
        for _ in 0..self.n_iters {
            // Generate y prediction
            let y_hat = X.dot(&w);

            // Compute gradient
            let residuals = &y_hat - y;
            let dw = (2.0 / n as f64) *X.t().dot(&residuals);

            w = w - self.lr*dw;
        }
        self.weights = Some(w);
    }

    pub fn predict(&mut self, X: &Array2<f64>) -> Array1<f64> {
        let w = self.weights.as_ref().expect(
            "Model has not been trained. Call .fit(X,y) first."
        );

        X.dot(w)
    }

    pub fn mse(&mut self, X: &Array2<f64>, y: &Array1<f64>) -> f64 {
        let y_hat = self.predict(X);
        let residuals = y - &y_hat;

        residuals.mapv(|err| err.powi(2)).mean().unwrap_or(0.0)
    }
}
